whelk-core/src/main/groovy/whelk/component/ElasticClient.groovy

```diff
@@ -0,0 +1,194 @@
+        // httpParams.setParameter(ClientPNames.CONN_MANAGER_TIMEOUT, new Long(TIMEOUT_MS));
+
+        return new ElasticClient(httpClient, elasticHosts)
+    }
+
+    ElasticClient(HttpClient httpClient, List<String> elasticHosts) {
+        this.httpClient = httpClient
+        this.elasticNodes = elasticHosts.collect { new ElasticNode(it) }
+
+        globalRetry = retryRegistry.retry(ElasticClient.class.getSimpleName(), RetryConfig.custom()
+                .waitDuration(Duration.ofMillis(10))
+                .maxAttempts(elasticNodes.size() * 2)
+                .build())
+
+        CollectorRegistry collectorRegistry = CollectorRegistry.defaultRegistry
+        collectorRegistry.register(RetryMetricsCollector.ofRetryRegistry(retryRegistry))
+        collectorRegistry.register(CircuitBreakerMetricsCollector.ofCircuitBreakerRegistry(circuitBreakerRegistry))
+
+        log.info "ElasticSearch component initialized with ${elasticHosts.size()} nodes."
+    }
+
+    Tuple2<Integer, String> performRequest(String method, String path, String body, String contentType0 = null) {
+        try {
+            def nodes = cycleNodes()
+            return globalRetry.executeSupplier({ -> nodes.next().performRequest(method, path, body, contentType0) })
+        }
+        catch (Exception e) {
+            log.warn("Request to ElasticSearch failed: ${e}", e)
+            throw new ElasticIOException(e.getMessage(), e)
+        }
+    }
+
+    private Iterator<ElasticNode> cycleNodes() {
+        def cycle = Iterators.cycle(elasticNodes)
+        Iterators.advance(cycle, random.nextInt(elasticNodes.size()))
+        return cycle
+    }
+
+    class ElasticNode {
+        String host
+        CircuitBreaker circuitBreaker
+        Retry retry
+
+        ElasticNode(String host) {
+            this.host = host
+            this.circuitBreaker = circuitBreakerRegistry.circuitBreaker(host, CB_CONFIG)
+            this.retry = retryRegistry.retry(host, RetryConfig.custom()
+                    .maxAttempts(2)
+                    .retryOnException({ !(it instanceof RetriesExceededException) }).build())
+        }
+
+        Tuple2<Integer, String> performRequest(String method, String path, String body, String contentType0 = null) {
+            HttpRequest request = buildRequest(method, path, body, contentType0)
+            return circuitBreaker.executeSupplier(Retry.decorateSupplier(retry, { -> sendRequest(request) }))
+        }
+
+        private Tuple2<Integer, String> sendRequest(HttpRequestBase request) {
+            try {
+                return sendRequestRetry429(request)
+            }
+            catch (Exception e) {
+                throw new RuntimeException(e.getMessage(), e)
+            }
+            finally {
+                request.reset()
+                request.releaseConnection()
+            }
+        }
+
+        private Tuple2<Integer, String> sendRequestRetry429(HttpRequestBase request) {
+            int backOffTime = 1
```

Comment:
Add unit to variable name

Reply:
:+1:

Reply:
Could use reslience4j backoff mechanism here too. Right now it's used moved from the old location.